File: old-projects/src/Monsters/Monster.ts
import { Point } from "../Point";
import { Sprite } from "../Sprite";
import { Game } from "../Game";
import { Util } from "../Util";

export abstract class Monster extends Sprite {
  protected life = 0;
  protected originalLife = 0;
  private static readonly size = 16;
  private static readonly halfSize = Monster.size / 2;
  private readonly path: Point[];
  private readonly maxSpeed: number;
  private speed: number;
  private damageIndicatorChanged = false;
  protected angle = 0;
  private xsign = 0;
  private ysign = 0;
  private currentDest: Point;
  protected damageAlpha = 0;

  protected constructor(x: number, y: number, travelPath: Point[], initialSpeed: number) {
    super();
    this.speed = this.maxSpeed = initialSpeed;
    this.path = travelPath.slice(2).map((point) => point.clone());
    this.x = x;
    this.y = y;
    this.currentDest = (travelPath[1] ?? travelPath[0]).clone();
    this.changeDirection(this.currentDest);
  }

  bounty = 0;
  color = "#fff";

  takeDamage(damage: number): void {
    this.life -= damage;
    if (this.life < 0) {
      this.life = 0;
    }
    if (this.damageAlpha === 0) {
      this.damageIndicatorChanged = true;
    }
    this.damageAlpha = 1;
  }

  slowDown(d: number): void {
    this.speed = this.maxSpeed * d;
  }

  override render(context: CanvasRenderingContext2D): void {
    const x = Math.round(this.x - Monster.halfSize);
    const length = Monster.size * this.life / this.originalLife;
    context.save();
    context.strokeStyle = "#008000";
    context.lineWidth = 2;
    context.beginPath();
    context.moveTo(x, Math.round(this.y - Monster.halfSize - 4));
    context.lineTo(Math.round(x + length), Math.round(this.y - Monster.halfSize - 4));
    context.stroke();
    context.restore();
  }

  override animate(): void {
    if (this.speed < this.maxSpeed) {
      this.speed += 0.01 * Game.currentFrameMultiplier;
      if (this.speed > this.maxSpeed) {
        this.speed = this.maxSpeed;
      }
      const velocity = Util.calculateLocation(this.angle, this.speed);
      this.dx = velocity.x;
      this.dy = velocity.y;
    }

    this.updatePosition();

    if (this.isOutsideBounds) {
      this.markedForRemoval = true;
      return;
    }

    if (this.life <= 0) {
      Game.killMonster(this);
      return;
    }

    const signChanged = Math.sign(this.currentDest.x - this.x) !== this.xsign || Math.sign(this.currentDest.y - this.y) !== this.ysign;
    if (signChanged) {
      if (this.path.length !== 0) {
        this.y = this.currentDest.y;
        this.x = this.currentDest.x;
        this.currentDest = this.path[0];
        this.path.splice(0, 1);
        this.changeDirection(this.currentDest);
      } else {
        Game.escapeMonster(this);
        return;
      }
    }

    if (this.damageAlpha > 0) {
      this.damageAlpha -= 0.03 * Game.currentFrameMultiplier;
      if (this.damageAlpha <= 0) {
        this.damageAlpha = 0;
        this.damageIndicatorChanged = true;
      }
    }

    this.damageIndicatorChanged = false;
  }

  protected angleChanged(newAngle: number): void {}

  private changeDirection(dest: Point): void {
    this.xsign = Math.sign(dest.x - this.x);
    this.ysign = Math.sign(dest.y - this.y);
    this.angle = Util.calculateAngle(this.x, this.y, dest.x, dest.y);
    this.angleChanged(this.angle);
    const velocity = Util.calculateLocation(this.angle, this.speed);
    this.dx = velocity.x;
    this.dy = velocity.y;
  }
}

File: silverlight-ports/vector-defence/original-csharp-source/Monsters/Monster.cs
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Media;
using System.Windows.Shapes;

namespace VectorDefenceSL.Monsters
{
    public abstract class Monster : Sprite
    {
        protected double life;
        protected double originalLife;

        private const double Size = 16;
        private const double HalfSize = Size / 2;
        private readonly List<Point> path;
        private readonly Line line;
        private readonly double maxSpeed;
        private double speed;
        private bool damangeIndicatorChanged;
        private double angle;
        private int xsign;
        private int ysign;
        private Point currentDest;
        private double damangeAlpha;

        protected Monster(double x, double y, IList<Point> travelPath, double initialSpeed)
        {
            line = new Line { StrokeThickness = 2, Stroke = new SolidColorBrush(Colors.Green) };
            Game.AddElement(line);
            speed = maxSpeed = initialSpeed;
            path = new List<Point>(travelPath);
            X = x;
            Y = y;
            currentDest = travelPath[0];
            ChangeDirection(currentDest);
        }

        public int Bounty { get; protected set; }

        public Color Color { get; protected set;  }

        public override void RemoveFromCanvas()
        {
            base.RemoveFromCanvas();
            Game.Children.Remove(line);
        }

        public void TakeDamage(double damage)
        {
            life -= damage;
            if (life < 0)
            {
                life = 0;
            }

            if (damangeAlpha == 0)
            {
                damangeIndicatorChanged = true;
            }
            damangeAlpha = 1;
        }

        public void SlowDown(double d)
        {
            speed = maxSpeed * d;
        }

        public override void Render()
        {
            double x = X - HalfSize;
            line.X1 = Math.Round(x);
            var length = Size * life / originalLife;
            line.X2 = Math.Round(x + length);
            line.Y1 = line.Y2 = Math.Round(Y - HalfSize - 4);

            if (damangeAlpha != 0)
            {
                mainElement.Fill.Opacity = damangeAlpha;
            }
        }

        public override void Animate()
        {
            if (speed < maxSpeed)
            {
                speed += .01 * Game.CurrentFrameMultiplier;
                if (speed > maxSpeed)
                {
                    speed = maxSpeed;
                }
                Util.CalculateLocation(angle, speed, out DX, out DY); // update dx and dy vectors
            }

            UpdatePosition();

            if (IsOutsideBounds)
            {
                MarkedForRemoval = true;
                return;
            }

            if (life <= 0)
            {
                Game.KillMonster(this);
                return;
            }

            bool signChanged = Math.Sign(currentDest.X - X) != xsign || Math.Sign(currentDest.Y - Y) != ysign;
            if (signChanged)
            {
                if (path.Count != 0)
                {
                    Y = currentDest.Y;
                    X = currentDest.X;

                    currentDest = path[0];
                    path.RemoveAt(0);
                    ChangeDirection(currentDest);
                }
                else
                {
                    Game.EscapeMonster(this);
                    return;
                }
            }

            if (damangeAlpha > 0)
            {
                damangeAlpha -= .03 * Game.CurrentFrameMultiplier;
                if (damangeAlpha <= 0)
                {
                    damangeAlpha = 0;
                    damangeIndicatorChanged = true;
                }
            }
            
            if (damangeIndicatorChanged)
            {
                mainElement.Fill = damangeAlpha == 0 ? new SolidColorBrush(Colors.Black) : new SolidColorBrush(Colors.Purple);
            }

            damangeIndicatorChanged = false;
        }

        protected virtual void AngleChanged(double newAngle)
        {
        }

        private void ChangeDirection(Point dest)
        {
            xsign = Math.Sign(dest.X - X);
            ysign = Math.Sign(dest.Y - Y);
            angle = Util.CalculateAngle(X, Y, dest.X, dest.Y);
            AngleChanged(angle);
            Util.CalculateLocation(angle, speed, out DX, out DY);
        }
    }
}
